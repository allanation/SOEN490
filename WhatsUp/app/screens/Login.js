/* eslint-disable react/no-unescaped-entities */
import React from "react";
import { useState } from "react";
import { StyleSheet, Text, View, Image, Alert } from "react-native";
import AppButton from "../components/AppButton";
import AppTextInput from "../components/AppTextInput";
import Links from "../components/Links";
import Screen from "../components/Screen";
import TitleHeaders from "../components/TitleHeaders";
import logo from "../Images/w3.png";
import { auth, db } from "../firebase";
import { collection, getDocs, query, where } from "firebase/firestore";
import { signInWithEmailAndPassword } from "firebase/auth";
import { useNavigation } from "@react-navigation/native";
import colors from "../config/colors";
import { SimpleLineIcons } from "@expo/vector-icons";

export default function Login() {
  const navigation = useNavigation();
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [isOrganizer] = useState(false);
  const handleLogin = async () => {
    await signInWithEmailAndPassword(auth, email, password)
      .then(async () => {
        const q = query(collection(db, "users"), where("email", "==", email));
        const querySnapshot = await getDocs(q);
        if (querySnapshot != null) {
          querySnapshot.forEach((doc) => {
            if (isOrganizer == doc.data().isOrganizer) {
              navigation.navigate("Attendee");
              setPassword("");
            } else {
              navigation.navigate("Organizer");
              setPassword("");
            }
          });
        }
      })
      .catch(() => {
        Alert.alert("Try again", "Invalid email or password.");
      });
  };
  const SignupPressed = async () => {
    navigation.navigate("SignUp");
  };

  const ResetPassword = () => {
    navigation.navigate("ResetPassword");
  };
  return (
    <Screen
      resizeMode='cover'
      style={{ paddingHorizontal: "10%" }}
    >
      <Image
        source={logo}
        style={{ marginTop: 18, width: 166, height: 212, alignSelf: "center" }}
      />
      <TitleHeaders
        style={{ fontSize: 38, marginTop: 20, marginBottom: 5 }}
        title='Login'
        isTitle={true}
      />
      <AppTextInput
        placeholder='Email'
        keyboardType='email-address'
        onChangeText={(text) => setEmail(text.toLowerCase())}
        value={email}
        style={{
          shadowColor: "black", // IOS
          shadowOffset: { height: 1, width: 1 }, // IOS
          shadowOpacity: 0.2, // IOS
          shadowRadius: 2, //IOS
          elevation: 4, // Android
        }}
      />
      <AppTextInput
        placeholder='Password'
        secureTextEntry
        onChangeText={(text) => setPassword(text)}
        value={password}
        style={{
          shadowColor: "black", // IOS
          shadowOffset: { height: 1, width: 1 }, // IOS
          shadowOpacity: 0.2, // IOS
          shadowRadius: 2, //IOS
          elevation: 4, // Android
        }}
      />
      <View style={styles.organizertwo}>
        <Links
          style={styles.link}
          link='Forgot Password?'
          onPress={ResetPassword}
        />
      </View>
      <AppButton title='Login' onPress={handleLogin} />
      <View style={{ flexDirection: "row", justifyContent: "center" }}>
        <Text style={styles.text}>Don't have an account? </Text>
        <Links style={styles.link} link='Sign up' onPress={SignupPressed} />
      </View>
      <View
        style={{
          flexDirection: "row",
          alignItems: "center",
          paddingVertical: 15,
        }}
      >
        <View style={{ flex: 1, height: 1, backgroundColor: "lightgrey" }} />
        <View>
          <Text
            style={{
              width: 105,
              textAlign: "center",
              color: "darkgray",
              fontSize: 14,
            }}
          >
            Or login with
          </Text>
        </View>
        <View style={{ flex: 1, height: 1, backgroundColor: "lightgrey" }} />
      </View>
      <View style={styles.rowContainer}>
        <SimpleLineIcons
          name='social-facebook'
          size={30}
          style={{
            color: colors.secondary,
            fontSize: 30,
            alignSelf: "center",
            paddingVertical: 10,
          }}
          onPress={() => console.log("Facebook")}
          testID='facebook'
        />
        <SimpleLineIcons
          name='social-twitter'
          size={30}
          style={{
            color: colors.secondary,
            fontSize: 30,
            paddingHorizontal: 40,
            alignSelf: "center",
          }}
          onPress={() => console.log("Facebook")}
          testID='twitter'
        />
        <SimpleLineIcons
          name='social-google'
          size={30}
          style={{
            color: colors.secondary,
            fontSize: 30,
            paddingVertical: 10,
          }}
          onPress={() => console.log("Facebook")}
          testID='google'
        />
      </View>
    </Screen>
  );
}

const styles = StyleSheet.create({
  organizertwo: {
    alignItems: "flex-end",
    width: "98%",
  },
  rowContainer: {
    flexDirection: "row",
    alignSelf: "center",
  },
  text: {
    color: colors.darkerGrey,
    fontSize: 16,
  },
});
