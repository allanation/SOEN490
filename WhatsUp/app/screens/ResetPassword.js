import React from "react";
import ImgOrgBottom from "../components/ImgOrgBottom"
import { useState } from "react";
import { StyleSheet, View, Alert } from "react-native";
import AppButton from "../components/AppButton";
import AppTextInput from "../components/AppTextInput";
import Links from "../components/Links";
import Screen from "../components/Screen";
import TitleHeaders from "../components/TitleHeaders";
import { useNavigation } from "@react-navigation/native";
import { auth } from "../firebase";
import { sendPasswordResetEmail } from "firebase/auth";
import { fetchSignInMethodsForEmail } from "firebase/auth";

export default function ResetPassword() {
  const navigation = useNavigation();
  const [userEmail, setEmail] = useState("");

  const sendResetEmail = () => {
    sendPasswordResetEmail(auth, userEmail)
      .then(() => {
        Alert.alert(
          "Email sent sucessfully.",
          "Check your emails to reset the password."
        );
      })
      .catch(() => {});
  };

  //Check if email exists in the database before sending the link
  const checkIfEmailExists = () => {
    fetchSignInMethodsForEmail(auth, userEmail)
      .then((result) => {
        if (result === undefined || result.length == 0) {
          Alert.alert("We couldn't find an account with that email address.");
        } else {
          //If email exists, send the link
          sendResetEmail();
        }
      })
      .catch((error) => {
        console.log(error.message);
        Alert.alert("Please enter a valid email address.");
      });
  };

  const Login = () => {
    navigation.navigate("Login");
  };

  return (
    <Screen style={{ padding: 10, marginTop: 80 }}>
      <TitleHeaders title="Forgot Password?" />
      <TitleHeaders isTitle = {false} title="Enter your email to receive a verification link" />
      <AppTextInput
        placeholder="Email"
        keyboardType="email-address"
        onChangeText={(text) => setEmail(text)}
        value={userEmail}
      />

      <AppButton title="Submit" onPress={checkIfEmailExists} />

      <View style={styles.organizertwo}>
        <Links style={styles.link} link="Back to Login" onPress={Login} />
      </View>
      <ImgOrgBottom resizeMode="contain" />
    </Screen>
  );
}

const styles = StyleSheet.create({
  organizertwo: {
    paddingLeft: 110,
    width: "98%",
  },
  link: {
    marginLeft: "14.1%",
  },
});
