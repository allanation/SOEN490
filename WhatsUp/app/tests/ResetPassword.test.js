import ResetPassword from "../screens/ResetPassword";
import "react-native";
import React from "react";
import {
  render,
  screen,
  fireEvent,
  waitFor,
} from "@testing-library/react-native";
import { Alert } from "react-native";
import { NavigationContainer } from "@react-navigation/native";
import Login from "../screens/Login";

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));

jest.useFakeTimers();
//jest.mock("../screens/ResetPassword.js");
// it("renders correctly", () => {
//   const tree = render(
//     <NavigationContainer>
//       <ResetPassword />
//     </NavigationContainer>
//   ).toJSON();
//   expect(tree).toMatchSnapshot();
// });

test("Reset password with a correct email", async () => {
  render(
    <NavigationContainer>
      <ResetPassword />
    </NavigationContainer>
  );
  await waitFor(() =>
    fireEvent.changeText(
      screen.getByPlaceholderText("Email"),
      "taversofiya@gmail.com"
    )
  );

  jest.spyOn(Alert, "alert");

  await waitFor(() => {
    fireEvent.press(screen.getByText("Submit"));
    expect(Alert.alert);
  });
});

test("Reset password with wrong email", async () => {
  render(
    <NavigationContainer>
      <ResetPassword />
    </NavigationContainer>
  );
  await waitFor(() =>
    fireEvent.changeText(
      screen.getByPlaceholderText("Email"),
      "notexistingemail@gmail.com"
    )
  );

  jest.spyOn(Alert, "alert");

  await waitFor(() => {
    fireEvent.press(screen.getByText("Submit"));
    expect(Alert.alert);
  });
});

test("Reset password with invalid email", async () => {
  render(
    <NavigationContainer>
      <ResetPassword />
    </NavigationContainer>
  );
  await waitFor(() =>
    fireEvent.changeText(screen.getByPlaceholderText("Email"), "")
  );

  jest.spyOn(Alert, "alert");

  await waitFor(() => {
    fireEvent.press(screen.getByText("Submit"));
    expect(Alert.alert);
  });
});

test("Go back to Login page", async () => {
  render(
    <NavigationContainer>
       <ResetPassword />
    </NavigationContainer>
  );

  await waitFor(() => fireEvent.press(screen.getByText("Back to Login")));

  render(
    <NavigationContainer>
      <Login />
    </NavigationContainer>
  );

  expect(screen.getAllByText("Login")).toBeTruthy();
});
