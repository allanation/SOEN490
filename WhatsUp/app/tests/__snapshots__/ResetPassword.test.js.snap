// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "flex": 1,
        "paddingTop": 20,
      },
      Object {
        "marginTop": 80,
        "padding": 10,
      },
    ]
  }
>
  <Image
    source={
      Object {
        "uri": undefined,
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Helvetica Neue",
          "marginLeft": "1%",
        },
        undefined,
        Object {
          "fontSize": 28,
          "fontWeight": "bold",
        },
      ]
    }
  >
    Forgot Password?
  </Text>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Helvetica Neue",
          "marginLeft": "1%",
        },
        undefined,
        Object {
          "color": "#747474",
          "fontSize": 20,
          "marginBottom": 6,
          "marginLeft": "1%",
          "paddingVertical": 6,
        },
      ]
    }
  >
    Enter your email to receive a verification link
  </Text>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "white",
          "borderColor": "#989797",
          "borderRadius": 12,
          "borderWidth": 1,
          "marginLeft": "1%",
          "marginVertical": 10,
          "opacity": 0.9,
          "paddingHorizontal": 12,
          "paddingVertical": 7,
          "width": "98%",
        },
        undefined,
      ]
    }
  >
    <TextInput
      keyboardType="email-address"
      onChangeText={[Function]}
      placeholder="Email"
      placeholderTextColor="grey"
      style={
        Object {
          "color": "black",
          "fontFamily": "Helvetica Neue",
          "fontSize": 18,
          "opacity": 1,
        }
      }
      value=""
    />
  </View>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#00c0a4",
        "borderRadius": 20,
        "elevation": 4,
        "marginLeft": "1%",
        "marginVertical": 40,
        "opacity": 1,
        "padding": 12,
        "shadowColor": "black",
        "shadowOffset": Object {
          "height": 1,
          "width": 1,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 3,
        "width": "90%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Helvetica Neue",
            "fontSize": 18,
            "fontWeight": "bold",
            "textAlign": "center",
            "textAlignVertical": "center",
          },
          undefined,
        ]
      }
    >
      Submit
    </Text>
  </View>
  <View
    style={
      Object {
        "paddingLeft": 110,
        "width": "98%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginLeft": "14.1%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#00c0a4",
            "fontSize": 16,
            "fontWeight": "400",
          }
        }
      >
        Back to Login
      </Text>
    </View>
  </View>
  <Image
    source={1}
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "bottom": 0,
          "height": "100%",
          "position": "absolute",
          "width": "100%",
        },
        Object {
          "bottom": 0,
          "height": 172.91414752116083,
          "position": "absolute",
          "resizeMode": "stretch",
          "width": 750,
        },
      ]
    }
  />
</View>
`;
